<a href="https://colab.research.google.com/github/pyatibratova/karpov.courses/blob/main/%D0%B7%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Выясните, какие пары товаров пользователи чаще всего покупают вместе. По сути, вам необходимо найти паттерны покупок, что позволит оптимизировать размещение продуктов в магазине, для удобства пользователей и увеличения выручки.**
      
id – означает покупку (в одну покупку входят все товары, купленные пользователем во время 1 похода в магазин)         
Товар – наименование товара      
Количество – число единиц купленного товара    
           
Воспользуйтесь этими данными и выясните, какие пары товаров пользователи чаще всего покупают вместе.        
        
Встречаемость – число раз, когда такая пара была встречена
Другими словами: 2 раза люди покупали одновременно чай и арбуз, 1 раз одновременно покупали арбуз и сгущёнку и 1 раз одновременно были куплены чай со сгущёнкой.           
          
Напишите код на python для получения нужной таблицы и укажите 5 наиболее распространённых паттернов.    

In [64]:
import pandas as pd
import itertools

url = "test1_completed.csv"
df = pd.read_csv(url)
df.head()

id            Товар  Количество
0  17119            Лимон         1.1
1  17119  Лимон оранжевый         0.7
2  17119        Лук-порей        10.0
3  17119     Лук репчатый         2.5
4  17119    Малина свежая         1.0

In [65]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43514 entries, 0 to 43513
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   id          43514 non-null  int64  
 1   Товар       43514 non-null  object 
 2   Количество  43514 non-null  float64
dtypes: float64(1), int64(1), object(1)
memory usage: 1020.0+ KB


In [66]:
df['Товар'].nunique()

199

В магазине предоставлено 199 уникальных товаров. Наша задача найти комбинации самых популярных пар и вывести их частоту покупок.

In [67]:
results = []
X = {}
for i in df.id.unique():
    product = df[df.id == i]['Товар'].to_list()
    product.sort()
    for j in list(itertools.combinations(product, 2)):
        X[j] = X.get(j, 0) + 1 

In [68]:
sort_x = sorted(X.items(), key=lambda x: x[1], reverse=True)

In [69]:
# преобразуем полученные данные в таблицу
data = pd.DataFrame.from_dict(sort_x)
data.columns = ['pair', 'count']
data.head()

pair  count
0    (Огурцы Луховицкие, Укроп)    431
1             (Петрушка, Укроп)    408
2    (Арбуз, Огурцы Луховицкие)    345
3  (Кабачки, Огурцы Луховицкие)    326
4                (Кинза, Укроп)    303

In [71]:
data['pair'] = [(str(val).replace('(', '').replace(')', '').replace('\'', '')) for val in data['pair']]
data.head()

pair  count
0    Огурцы Луховицкие, Укроп    431
1             Петрушка, Укроп    408
2    Арбуз, Огурцы Луховицкие    345
3  Кабачки, Огурцы Луховицкие    326
4                Кинза, Укроп    303

In [77]:
pattern = data['pair'].str.split(',',expand=True)
pattern["count"] = data["count"] 
pattern = pattern.drop(pattern.columns[[2, 3] ],axis=1)
pattern.columns = ['Товар_1', 'Товар_2', 'Встречаемость']
pattern.head(5)

Товар_1             Товар_2  Встречаемость
0  Огурцы Луховицкие               Укроп            431
1           Петрушка               Укроп            408
2              Арбуз   Огурцы Луховицкие            345
3            Кабачки   Огурцы Луховицкие            326
4              Кинза               Укроп            303

**Вывод**         
Нам не известно, за какой период времени получены данные о покупках. 
Поэтому сделаем предположение, что в период июль - сентябрь.
Тогда понятно, почему люди покупают много зелени, огурцов и арбузы.